art/preprocessing/audio/l_filter/numpy.py

```diff
@@ -108,6 +108,87 @@ def __call__(self, x: np.ndarray, y: Optional[np.ndarray] = None) -> Tuple[np.nd
 
         return x_preprocess, y
 
+    def estimate_gradient(self, x: np.ndarray, grad: np.ndarray) -> np.ndarray:
+        """
+        Provide an estimate of the gradients of the defence for the backward pass.
+
+        :param x: Input data for which the gradient is estimated. First dimension is the batch size.
+        :param grad: Gradient value so far.
+        :return: The gradient (estimate) of the defence.
+        """
+        if (grad.shape != x.shape) and (
+            (len(grad.shape) != len(x.shape) + 1) or (grad.shape[2:] != x.shape[1:]) or (grad.shape[0] != x.shape[0])
+        ):
+            raise ValueError(
+                "The `so far` gradient shape {} does not suit the input shape {}".format(grad.shape, x.shape)
+            )
+
+        if not (self.denominator_coef[0] == 1.0 and np.sum(self.denominator_coef) == 1.0):
+            logger.warning(
+                "Gradient estimation for a non-fir filtering operation is not supported. A BPDA gradient estimation "
+                "is returned."
+            )
+            return grad
+
+        # We will compute gradient for one sample at a time
+        x_grad_list = list()
+
+        if grad.shape == x.shape:
+            for i, x_i in enumerate(x):
+                # First compute the gradient matrix
+                grad_matrix = self._compute_gradient_matrix(size=x_i.size)
+
+                # Then combine with the input gradient
+                grad_x_i = np.zeros(x_i.size)
+                for j in range(x_i.size):
+                    grad_x_i[j] = np.sum(grad[i] * grad_matrix[:, j])
+
+                # And store the result
+                x_grad_list.append(grad_x_i)
+
+        else:
+            for i, x_i in enumerate(x):
+                # First compute the gradient matrix
+                grad_matrix = self._compute_gradient_matrix(size=x_i.size)
+
+                # Then combine with the input gradient
+                grad_x_i = np.zeros_like(grad[i])
+                for j in range(x_i.size):
+                    grad_x_i[:, j] = np.sum(grad[i, :, :] * grad_matrix[:, j], axis=1)
+
+                # And store the result
+                x_grad_list.append(grad_x_i)
```

Comment:
The two outer for-loops are very similar except of their inner for-loops. Could we combine the two outer for-loops by moving the if-condition into the loop and reduce the number of identical lines?

Reply:
@beat-buesser : Yes I thought about this, then I decided this form because I don't want the if-condition to be called repeatedly and unnecessarily in the for loop. But it will a bit slower but that's ok to move the if-condition inside.